UpgradeTest.upgradeWithoutSnapshot "flaky" - or at least not robust enough

**Summary**

- How often does the test fail? Seen once so far
- Does it block your work? yes
- Do we suspect that it is a real failure? no

**Failures**

https://ci.zeebe.camunda.cloud/job/zeebe-io/job/zeebe/job/4340-do-not-merge-monitor-build-health-branch/164/testReport/

The test failed due to an internal server error of a dependent resource.

<details><summary>Example assertion failure</summary>
<pre>
Error Message

Container startup failed 

Stacktrace

org.testcontainers.containers.ContainerLaunchException: Container startup failed
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:320)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:300)
	at io.zeebe.test.ContainerStateRule.start(ContainerStateRule.java:80)
	at io.zeebe.test.UpgradeTest.upgradeZeebe(UpgradeTest.java:203)
	at io.zeebe.test.UpgradeTest.upgradeWithoutSnapshot(UpgradeTest.java:198)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=camunda/zeebe:0.23.0, imagePullPolicy=DefaultPullPolicy())
	at org.testcontainers.containers.GenericContainer.getDockerImageName(GenericContainer.java:1268)
	at org.testcontainers.containers.GenericContainer.logger(GenericContainer.java:603)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:309)
	... 18 more
Caused by: com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"Get https://quay.io/v2/testcontainers/ryuk/manifests/0.2.3: received unexpected HTTP status: 500 Internal Server Error"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:293)
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.lambda$executeAndStream$4(OkHttpInvocationBuilder.java:317)
	... 1 more

</pre>
</details>

**Hypotheses**

**Logs**


<details><summary>Logs</summary>
<pre>
01:15:30.990 [main] [] INFO  io.zeebe.test - Test started: upgradeWithoutSnapshot[job](io.zeebe.test.UpgradeTest)
01:15:34.817 [Time-limited test] [] INFO  org.testcontainers.dockerclient.EnvironmentAndSystemPropertyClientProviderStrategy - Found docker client settings from environment
01:15:34.887 [Time-limited test] [] INFO  org.testcontainers.dockerclient.DockerClientProviderStrategy - Found Docker environment with Environment variables, system properties and defaults. Resolved dockerHost=tcp://localhost:2375
01:15:35.813 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - Docker host IP address is localhost
01:15:35.993 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - Connected to docker: 
  Server Version: 18.09.4
  API Version: 1.39
  Operating System: Alpine Linux v3.9 (containerized)
  Total Memory: 120871 MB
01:15:36.110 [Time-limited test] [] WARN  org.testcontainers.utility.RegistryAuthLocator - Failure when attempting to lookup auth config (dockerImageName: quay.io/testcontainers/ryuk:0.2.3, configFile: /root/.docker/config.json. Falling back to docker-java default behaviour. Exception message: /root/.docker/config.json (No such file or directory)
01:15:36.990 [tc-okhttp-stream-2050240984] [] ERROR com.github.dockerjava.api.async.ResultCallbackTemplate - Error during callback
com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"Get https://quay.io/v2/testcontainers/ryuk/manifests/0.2.3: received unexpected HTTP status: 500 Internal Server Error"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:293) ~[testcontainers-1.14.1.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.lambda$executeAndStream$4(OkHttpInvocationBuilder.java:317) ~[testcontainers-1.14.1.jar:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
01:15:37.002 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - Docker host IP address is localhost
01:15:37.086 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - Connected to docker: 
  Server Version: 18.09.4
  API Version: 1.39
  Operating System: Alpine Linux v3.9 (containerized)
  Total Memory: 120871 MB
01:15:38.303 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Starting to pull image
01:15:38.399 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  0 pending,  0 downloaded,  0 extracted, (0 bytes/0 bytes)
01:15:39.186 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  2 pending,  1 downloaded,  0 extracted, (527 KB/? MB)
01:15:44.488 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  1 pending,  2 downloaded,  0 extracted, (559 KB/? MB)
01:15:44.695 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  1 pending,  2 downloaded,  1 extracted, (3 MB/? MB)
01:15:44.798 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  0 pending,  3 downloaded,  1 extracted, (3 MB/2 MB)
01:15:44.988 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  0 pending,  3 downloaded,  2 extracted, (3 MB/2 MB)
01:15:45.093 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pulling image layers:  0 pending,  3 downloaded,  3 extracted, (4 MB/2 MB)
01:15:45.191 [tc-okhttp-stream-1195649370] [] INFO  org.testcontainers.DockerClientFactory - Pull complete. 3 layers, pulled in 6s (downloaded 4 MB at 786 KB/s)
01:15:48.016 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - Ryuk started - will monitor and terminate Testcontainers containers on JVM exit
01:15:48.016 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - Checking the system...
01:15:48.017 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - ✔︎ Docker server version should be at least 1.6.0
01:15:48.487 [Time-limited test] [] INFO  org.testcontainers.DockerClientFactory - ✔︎ Docker environment should have more than 2GB free disk space
01:15:48.489 [Time-limited test] [] ERROR io.zeebe.test.ContainerStateRule - 
===============================================
Broker logs
===============================================

01:15:48.496 [main] [] ERROR io.zeebe.test.ContainerStateRule - 
===============================================
Broker logs
===============================================
</pre>
</details>

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -91,6 +91,8 @@
     <version.config>1.4.0</version.config>
     <version.kryo>4.0.2</version.kryo>
     <version.awaitility>4.0.3</version.awaitility>
+    <version.failsafe>2.4.0</version.failsafe>
+
 
     <!-- maven plugins -->
     <plugin.version.antrun>3.0.0</plugin.version.antrun>
@@ -656,6 +658,12 @@
         <artifactId>classgraph</artifactId>
         <groupId>io.github.classgraph</groupId>
         <version>${version.classgraph}</version>
+      </dependency>
+
+      <dependency>
+        <groupId>net.jodah</groupId>
+        <artifactId>failsafe</artifactId>
+        <version>${version.failsafe}</version>
       </dependency>
 
     </dependencies>

--- a/upgrade-tests/pom.xml
+++ b/upgrade-tests/pom.xml
@@ -1,4 +1,6 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
 
   <modelVersion>4.0.0</modelVersion>
   <name>Zeebe Upgradability Tests</name>
@@ -58,6 +60,10 @@
     <dependency>
       <groupId>org.awaitility</groupId>
       <artifactId>awaitility</artifactId>
+    </dependency>
+    <dependency>    
+      <groupId>net.jodah</groupId>
+      <artifactId>failsafe</artifactId>
       <scope>test</scope>
     </dependency>
   </dependencies>

--- a/upgrade-tests/src/test/java/io/zeebe/test/ContainerStateRule.java
+++ b/upgrade-tests/src/test/java/io/zeebe/test/ContainerStateRule.java
@@ -12,21 +12,43 @@ import io.zeebe.containers.ZeebeBrokerContainer;
 import io.zeebe.containers.ZeebePort;
 import io.zeebe.containers.ZeebeStandaloneGatewayContainer;
 import java.time.Duration;
+import java.time.temporal.ChronoUnit;
 import java.util.Arrays;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.RetryPolicy;
 import org.junit.rules.TestWatcher;
 import org.junit.runner.Description;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.slf4j.event.Level;
+import org.testcontainers.containers.ContainerFetchException;
+import org.testcontainers.containers.ContainerLaunchException;
 import org.testcontainers.containers.Network;
 
 class ContainerStateRule extends TestWatcher {
 
+  private static final RetryPolicy<Void> CONTAINER_START_RETRY_POLICY =
+      new RetryPolicy().withMaxRetries(5).withBackoff(3, 30, ChronoUnit.SECONDS);
+
   private static final Pattern DOUBLE_NEWLINE = Pattern.compile("\n\n");
   private static final Duration CLOSE_TIMEOUT = Duration.ofSeconds(30);
   private static final Logger LOG = LoggerFactory.getLogger(ContainerStateRule.class);
+
+  static {
+    CONTAINER_START_RETRY_POLICY
+        .handleIf(
+            error ->
+                error instanceof ContainerLaunchException
+                    && error.getCause() instanceof ContainerFetchException)
+        .onRetry(
+            event -> {
+              LOG.info("Attempt " + event.getAttemptCount());
+              LOG.info("Retrying container start after exception:", event.getLastFailure());
+            });
+  }
+
   private ZeebeBrokerContainer broker;
   private ZeebeStandaloneGatewayContainer gateway;
   private ZeebeClient client;
@@ -79,7 +101,8 @@ class ContainerStateRule extends TestWatcher {
             .withLogLevel(Level.DEBUG)
             .withDebug(false);
 
-    broker.start();
+    Failsafe.with(CONTAINER_START_RETRY_POLICY).run(() -> broker.start());
+
     String contactPoint = broker.getExternalAddress(ZeebePort.GATEWAY);
 
     if (gatewayVersion != null) {
@@ -89,7 +112,8 @@ class ContainerStateRule extends TestWatcher {
               .withEnv("ZEEBE_GATEWAY_CLUSTER_CLUSTERNAME", "zeebe-cluster")
               .withNetwork(network)
               .withLogLevel(Level.DEBUG);
-      gateway.start();
+
+      Failsafe.with(CONTAINER_START_RETRY_POLICY).run(() -> gateway.start());
 
       contactPoint = gateway.getExternalAddress(ZeebePort.GATEWAY);
     }

